Dockerfile

```diff
@@ -0,0 +1,60 @@
+# Operating system with basic tools
+FROM --platform=linux/amd64 ubuntu:20.04 as base
+SHELL ["bash", "-c"]
+ENV TZ=UTC
+RUN ln -snf /usr/share/zoneinfo/$TZ /etc/localtime && echo $TZ > /etc/timezone && \
+    apt -yq update && \
+    apt -yqq install --no-install-recommends curl ca-certificates \
+        build-essential pkg-config libssl-dev llvm-dev liblmdb-dev clang cmake \
+        git jq npm xxd file curl unzip
+
+# Code specific dependencies
+FROM base as builder
+WORKDIR /code
+ENV RUSTUP_HOME=/opt/rustup \
+    CARGO_HOME=/opt/cargo \
+    FNM_DIR=/opt/fnm
+# Path modifications need to be done in separate ENV statements
+ENV PATH=$CARGO_HOME/bin:$PATH
+ENV PATH=$FNM_DIR/bin:$PATH
+# Install Rust and the Node.js version manager
+COPY rust-toolchain.toml .
+COPY .nvmrc .
+RUN curl -fsSL https://sh.rustup.rs -sSf | sh -s -- -y --no-modify-path && \
+    curl -fsSL https://fnm.vercel.app/install | bash -s -- --install-dir $FNM_DIR/bin --skip-shell
+# Add the fnm env var envaluation to the bashrc to enable it in bash by default
+RUN echo "eval \"$(fnm env)\"" >> $HOME/.bashrc
+# Install the correct version of Node.js and the dependencies
+COPY . .
+RUN eval "$(fnm env)" && \
```

Comment:
IIRC this line will run again if Docker detects any changes in any of the previous lines, and since `COPY . .` copies all files, any changes to the source files will result in all of this stuff being run again.

So maybe it would be more cache effecient to split it.

```
COPY .nvmrc .
RUN eval "$(fnm env)" && \
    fnm install && \
    fnm use

COPY package.json .
RUN corepack enable && \
    pnpm install

COPY . .
```

Reply:
> COPY . . copies all files

there's already a .dockerignore but still more files are copied than needed

Reply:
I made a small adjustment that improves a bit the caching of the layers, but pnpm install needs to happen after the `COPY . .` because it needs the package json from the inner monorepo projects

Reply:
the js environment also needs some source code in a postinstall hook that `pnpm install` triggers